README.md
# MemoryTools

[![PSGallery Version](https://img.shields.io/powershellgallery/v/MemoryTools.png?style=for-the-badge&logo=powershell&label=PowerShell%20Gallery)](https://www.powershellgallery.com/packages/MemoryTools/) [![PSGallery Downloads](https://img.shields.io/powershellgallery/dt/MemoryTools.png?style=for-the-badge&label=Downloads)](https://www.powershellgallery.com/packages/MemoryTools/)

This module contains a set of PowerShell functions for reporting on computer memory utilization and configuration. The commands use `Get-CimInstance`, so PowerShell 3.0 or later must be installed on remote computers, and they need to be configured for PowerShell remoting.

The project was first described on my [blog](http://bit.ly/1Tooj3Q).

Install the latest version of this module from the PowerShell Gallery. It should work in Windows PowerShell 5.1 and PowerShell 7 on a Windows platform.

```powershell
Install-Module MemoryTools [-scope CurrentUser]
```

## [Get-MemoryUsage](docs/Get-MemoryUsage.md)

This command will write a custom memory utilization object to the pipeline that indicates the current memory state.

```powershell
PS C:\> Get-MemoryUsage

Computername Status PctFree FreeGB TotalGB
------------ ------ ------- ------ -------
PROSPERO     OK        73.5  46.83      64
```

## [Show-MemoryUsage](docs/Show-MemoryUsage.md)

This command will also get the same information as Get-MemoryUsage but will display it with colorized output.

![Alt Colorized Memory Usage](assets/show-memoryusage.png)

## [Test-MemoryUsage](docs/Test-MemoryUsage.md)

This command can be used to test if memory utilization meets some criteria. There are several parameter sets for different tests. All of them can be used with `-Quiet` to return a simple Boolean value.

### Percent Free

The default behavior is to see if there is at least 50% free memory, but you can specify a different value.

```powershell
PS C:\> Test-MemoryUsage

Computername PctFree Test
------------ ------- ----
PROSPERO       73.23 True

```

### FreeGB

Test if there is at least X amount of free memory available.

```powershell
PS C:\> Test-MemoryUsage -freegb 8

Computername FreeGB Test
------------ ------ ----
PROSPERO      46.66 True

```

### TotalGB

Test if the computer has at least X amount of total memory.

```powershell
PS C:\> Test-MemoryUsage -TotalGB 16

Computername TotalGB Test
------------ ------- ----
PROSPERO          64 True

```

## UsedGB

Test if the computer is using X amount of memory or greater.

```powershell
PS C:\> Test-MemoryUsage -UsedGB 4

Computername UsedGB Test
------------ ------ ----
PROSPERO      17.37 True

```

## [Get-MemoryPerformance](docs/Get-MemoryPerformance.md)

This command will query memory performance counters.

```powershell
PS C:\> Get-MemoryPerformance

AvailableBytes                       : 36180754432
AvailableKBytes                      : 35332768
AvailableMBytes                      : 34504
CacheBytes                           : 436342784
CacheBytesPeak                       : 588001280
CacheFaultsPersec                    : 0
CommitLimit                          : 78617718784
CommittedBytes                       : 40527859712
DemandZeroFaultsPersec               : 251
FreeAndZeroPageListBytes             : 20220755968
FreeSystemPageTableEntries           : 12448913
LongTermAverageStandbyCacheLifetimes : 14400
ModifiedPageListBytes                : 1235890176
PageFaultsPersec                     : 232
PageReadsPersec                      : 0
PagesInputPersec                     : 0
PagesOutputPersec                    : 0
PagesPersec                          : 0
PageWritesPersec                     : 0
PercentCommittedBytesInUse           : 51
PoolNonpagedAllocs                   : 0
PoolNonpagedBytes                    : 1043083264
PoolPagedAllocs                      : 0
PoolPagedBytes                       : 858091520
PoolPagedResidentBytes               : 802910208
StandbyCacheCoreBytes                : 128360448
StandbyCacheNormalPriorityBytes      : 6915219456
StandbyCacheReserveBytes             : 8916418560
SystemCacheResidentBytes             : 436342784
SystemCodeResidentBytes              : 8192
SystemCodeTotalBytes                 : 8192
SystemDriverResidentBytes            : 15392768
SystemDriverTotalBytes               : 40984576
TransitionFaultsPersec               : 11
TransitionPagesRePurposedPersec      : 0
WriteCopiesPersec                    : 0
DateTime                             : 11/11/2020 5:56:08 PM
ComputerName                         : PROSPERO
```

Or you might use it like this:

```powershell
PS C:\> Get-CimSession | Get-MemoryPerformance | Select-Object Computername,CommittedBytes,AvailableMBytes,CacheBytes

ComputerName CommittedBytes AvailableMBytes CacheBytes
------------ -------------- --------------- ----------
thinkp1          4630245376           29007  139804672
win10            1045913600             989   69541888
srv1              598110208             592   22425600
dom1             1357385728            2913   52137984
srv2              586850304             534   28585984
```

## [Get-PhysicalMemory](docs/Get-PhysicalMemory.md)

This command will query the Win32_PhysicalMemory class to get hardware details.

```powershell
PS C:\> Get-PhysicalMemory -Computername thinkp1

   Computername: THINKP1

Manufacturer CapacityGB Form   ClockSpeed Voltage Location
------------ ---------- ----   ---------- ------- --------
Samsung      16         SODIMM 2667       1200    ChannelA-DIMM0
Micron       16         SODIMM 2667       1200    ChannelB-DIMM0
```

## [Get-TopProcessMemory](docs/Get-TopProcessMemory.md)

This command will use `Get-CimInstance` to retrieve the top processes by *Workingset*. It will retrieve the top 5 by default. The output will include a percentage of the total memory in use and the process owner.

```powershell
PS C:\> Get-TopProcessMemory

   Computername: PROSPERO

ProcessID   : 4280
Name        : Memory Compression
WS(MB)      : 1306.37
PctUsed     : 4.9505
Owner       : NT AUTHORITY\SYSTEM
Commandline :

ProcessID   : 40800
Name        : pwsh.exe
WS(MB)      : 1250.1
PctUsed     : 4.7373
Owner       : PROSPERO\Jeff
Commandline : "C:\Program Files\PowerShell\7\pwsh.exe" -nologo

ProcessID   : 43612
Name        : pwsh.exe
WS(MB)      : 534.37
PctUsed     : 2.025
Owner       : PROSPERO\Jeff
Commandline : "C:\Program Files\PowerShell\7\pwsh.exe" -NoProfile
              -ExecutionPolicy Bypass -Command "Import-Module 'c:\Users\Jeff\.vs
              code\extensions\ms-vscode.powershell-2023.2.1\modules\PowerShellEd
              itorServices\PowerShellEditorServices.psd1'; Start-EditorServices
              -HostName 'Visual Studio Code Host' -HostProfileId
              'Microsoft.VSCode' -HostVersion '2023.2.1' -AdditionalModules
              @('PowerShellEditorServices.VSCode') -BundledModulesPath 'c:\Users
              \Jeff\.vscode\extensions\ms-vscode.powershell-2023.2.1\modules'
              -EnableConsoleRepl -StartupBanner '' -LogLevel 'Normal' -LogPath '
              c:\Users\Jeff\AppData\Roaming\Code\User\globalStorage\ms-vscode.po
              wershell\logs\1679663113-13d77409-f0b9-4af6-87e8-58d6de36834f16796
              62816963\EditorServices.log' -SessionDetailsPath 'c:\Users\Jeff\Ap
              pData\Roaming\Code\User\globalStorage\ms-vscode.powershell\session
              s\PSES-VSCode-29600-898846.json' -FeatureFlags @() "

ProcessID   : 13480
Name        : PhoneExperienceHost.exe
WS(MB)      : 431.92
PctUsed     : 1.6368
Owner       : PROSPERO\Jeff
Commandline : "C:\Program Files\WindowsApps\Microsoft.YourPhone_1.23022.139.0_x6
              4__8wekyb3d8bbwe\PhoneExperienceHost.exe"
              -Restart:{F534506B-042E-4975-BF2C-806D54F0EBC7}

ProcessID   : 31184
Name        : firefox.exe
WS(MB)      : 349.93
PctUsed     : 1.3261
Owner       : PROSPERO\Jeff
Commandline : "C:\Program Files\Mozilla Firefox\firefox.exe" -osint -url
              https://pester.dev/docs/commands/Should
```

## [Get-ProcessMemory](docs/Get-ProcessMemory.md)

Get a snapshot of a process' memory usage based on its workingset value. Processes are grouped by name.

```powershell
PS C:\> Get-ProcessMemory | Select-Object -first 5

Name                   Count Threads      AvgMB      SumMB     Computername
----                   ----- -------      -----      -----     ------------
firefox                   17     467   142.2585  2418.3945         PROSPERO
pwsh                       3     101   642.0807  1926.2422         PROSPERO
svchost                  102     692    16.0955  1641.7461         PROSPERO
Code                      13     264    117.466  1527.0586         PROSPERO
Memory Compression         1      54  1306.3555  1306.3555         PROSPERO
```

## Roadmap

I have plans to revise commands to take advantage of SSH remoting in PowerShell 7. It also might be nice to have a WPF-based GUI to display memory information.


changelog.md
# Memory Tools Changelog

## [Unreleased]

## [1.1.0] - 2023-03-24

### Added

- Added `Get-MemoryUsage`.
- Added format file `myProcessMemoryUsage.format.ps1xml`.
- Help updates.
- Updated `README.md`.

### Changed

- General code clean up.
- Restructured module, separating functions to separate files.
- Migrated changelog to [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) format.
- Moved pre-release change log information to `pre-release-changelog.md`.

## [1.0.0] - 2020-11-11

### Added

- Added a custom format file for the `physicalMemoryUnit` object type.
- Added a custom format file for the `topProcessMemoryUnit` object type.
- Added online help links.

### Changed

- Modified `Show-MemoryUsage` to use ANSI-escape sequences instead of `Write-Host`.
- Modified `Get-PhysicalMemory` to write a custom object called `physicalMemoryUnit` to the pipeline.
- Modified `Get-TopProcessMemory` to write a custom object called `topProcessMemoryUnit` to the pipeline.
- Updated `README.md` and `license.txt`.
- Code update and cleanup for PowerShell 7 compatibility.
- Modified manifest to require PowerShell 5.1 as the minimum and support Core.
- Modified table layout in `MyMemoryUsage.format.ps1xml`.
- Updated Pester tests.

[Unreleased]: https://github.com/jdhitsolutions/MemoryTools/compare/v1.1.0..HEAD
[1.1.0]: https://github.com/jdhitsolutions/MemoryTools/compare/v1.0.0..v1.1.0
[1.0.0]: https://github.com/jdhitsolutions/MemoryTools/tree/v1.0.0

functions/public.ps1

Function Get-MemoryUsage {
    [cmdletbinding(DefaultParameterSetName = 'Computername')]
    [OutputType("MyMemoryUsage")]
    [alias("gmem")]
    Param(
        [Parameter(
            Position = 0,
            ValueFromPipeline,
            ValueFromPipelineByPropertyName,
            ParameterSetName = 'Computername'
        )]
        [ValidateNotNullorEmpty()]
        [Alias("cn")]
        [string[]]$Computername = $env:Computername,

        [ValidateSet("All", "OK", "Warning", "Critical")]
        [string]$Status = "All",

        [Parameter(ParameterSetName = 'Cim', Mandatory, ValueFromPipeline)]
        [ValidateNotNullorEmpty()]
        [Microsoft.Management.Infrastructure.CimSession[]]$CimSession
    )

    Begin {
        Write-Verbose "[BEGIN  ] Starting: $($MyInvocation.Mycommand)"
        $MyCimSession = @()
    } #begin

    Process {
        Write-Verbose "[PROCESS] PSBoundParameters"
        Write-Verbose ($PSBoundParameters | Out-String)

        Write-Verbose "[PROCESS] Using parameter set $($PSCmdlet.ParameterSetName)"

        if ($pscmdlet.ParameterSetName -eq 'Computername') {
            #create a temporary cimsession if using a computername
            foreach ($item in $Computername) {
                Try {
                    Write-Verbose "[PROCESS] Creating temporary CIM Session to $item"
                    $MyCIMSession += New-CimSession -ComputerName $item -ErrorAction Stop -OutVariable +tmpcs
                    $tmpSession = $True
                    Write-Verbose "[PROCESS] Added session"
                }
                Catch {
                    Write-Error "[$($item.toUpper())] Failed to create temporary CIM Session. $($_.exception.message)"
                }
            } #foreach item in computername
        } #if computername parameter set
        else {
            Write-Verbose "[PROCESS] Re-using CimSessions"
            $MyCIMSession = $CimSession
        }

        foreach ($session in $MyCIMSession) {

            Write-Verbose "[PROCESS] Processing $($session.computername) with session ID $($session.ID)"

            Try {
                $os = Get-CimInstance -ClassName Win32_OperatingSystem -CimSession $session -ErrorAction stop
            }
            Catch {
                Write-Error "[$($session.Computername.toUpper())] Failed to retrieve data. $($_.exception.message)"
            }
            if ($os) {
                # Determine if Dynamic Memory is used
                $MaxDynamicMemory = (Get-Counter -Counter "\Hyper-V Dynamic Memory Integration Service\Maximum Memory, Mbytes" -ComputerName $os.PSComputerName -ErrorAction SilentlyContinue).CounterSamples.CookedValue * 1KB

                # Determine the amount of free memory
                if ($MaxDynamicMemory) {
                    $FreeMemory = $os.FreePhysicalMemory + ($MaxDynamicMemory - $os.TotalVisibleMemorySize)
                    $TotalMemory = $MaxDynamicMemory
                }
                else {
                    $FreeMemory = $os.FreePhysicalMemory
                    $TotalMemory = $os.TotalVisibleMemorySize
                }

                $pctFree = [math]::Round(($FreeMemory / $TotalMemory) * 100, 2)

                if ($pctFree -ge $MemoryToolsOK) {
                    $StatusProperty = "OK"
                }
                elseif ($pctFree -ge $MemoryToolsWarning ) {
                    $StatusProperty = "Warning"
                }
                else {
                    #anything else is considered critical
                    $StatusProperty = "Critical"
                }

                $obj = [PSCustomObject]@{
                    PSTypename   = "MyMemoryUsage"
                    Computername = $OS.PSComputername.ToUpper()
                    Status       = $StatusProperty
                    PctFree      = $pctFree
                    FreeGB       = [math]::Round($FreeMemory / 1mb, 2)
                    TotalGB      = [int]($TotalMemory / 1mb)
                }

                #write object to the pipeline
                if ($Status -eq 'All') {
                    $obj
                }
                else {
                    #write filtered results
                    $obj | Where-Object { $_.Status -match $Status }
                }
                #reset variables erring on the side of caution
                Clear-Variable OS, obj, Mycimsession

            } #if OS

        } #foreach

        #clean up temporary sessions
        if ($tmpSession) {
            Write-Verbose "[PROCESS] Removing temporary sessions"
            $tmpcs | Out-String | Write-Verbose
            $tmpcs | Remove-CimSession
            Remove-Variable -name tmpcs
        }
    } #process

    End {
        Write-Verbose "[END    ] $($MyInvocation.Mycommand)"
    } #end

} #end Get-MemoryUsage

Function Show-MemoryUsage {

    [cmdletbinding(DefaultParameterSetName = 'Computername')]
    [alias("shmem")]
    Param(
        [Parameter(
            Position = 0,
            ValueFromPipeline,
            ValueFromPipelineByPropertyName,
            ParameterSetName = 'Computername'
        )]
        [ValidateNotNullorEmpty()]
        [Alias("cn")]
        [string[]]$Computername = $env:Computername,

        [Parameter(ParameterSetName = 'Cim', Mandatory, ValueFromPipeline)]
        [ValidateNotNullorEmpty()]
        [Microsoft.Management.Infrastructure.CimSession[]]$CimSession
    )

    Begin {
        Write-Verbose "[BEGIN  ] Starting: $($MyInvocation.Mycommand)"

        #a formatted report title
        $title = @"

    ****************
      Memory Check
    ****************
"@

        "$([char]0x1b)[38;5;159m$Title$([char]0x1b)[0m"
        # Write-Host $title -ForegroundColor Cyan

        #initialize an array to hold data
        $data = @()
    } #begin

    Process {
        Write-Verbose "[PROCESS] PSBoundParameters"
        Write-Verbose ($PSBoundParameters | Out-String)

        Write-Verbose "[PROCESS] Using parameter set $($PSCmdlet.ParameterSetName)"

        if ($pscmdlet.ParameterSetName -eq 'Computername') {
            foreach ($Computer in $Computername) {
                #get memory usage data for each computer
                $data += Get-MemoryUsage -Computername $computer
            }
        }
        else {
            foreach ($session in $CIMSession) {
                #get memory usage data for each computer
                $data += Get-MemoryUsage -CimSession $session
            }
        }
    } #Process

    End {
        #format the results using a custom formatting view
        $data | Format-Table -View show
        Write-Verbose "[END    ] $($MyInvocation.Mycommand)"
    } #end

} #end Show-MemoryUsage

#get-memory usage and test for a minimum %free, FreeGB, TotalGB or UsedGB
Function Test-MemoryUsage {
    [cmdletbinding(DefaultParameterSetName = 'PercentComputer')]
    [alias("tmem")]
    [OutputType("Boolean", "PSCustomObject")]
    Param(
        [ValidateNotNullorEmpty()]
        [Alias("cn")]
        [Parameter(Position = 0, ValueFromPipeline, ValueFromPipelineByPropertyName, ParameterSetName = "PercentComputer")]
        [Parameter(Position = 0, ValueFromPipeline, ValueFromPipelineByPropertyName, ParameterSetName = "FreeComputer")]
        [Parameter(Position = 0, ValueFromPipeline, ValueFromPipelineByPropertyName, ParameterSetName = "TotalComputer")]
        [Parameter(Position = 0, ValueFromPipeline, ValueFromPipelineByPropertyName, ParameterSetName = "UsedComputer")]
        [string[]]$Computername = $env:Computername,

        [Parameter(Mandatory, ParameterSetName = "PercentCIM")]
        [Parameter(Mandatory, ParameterSetName = "FreeCIM")]
        [Parameter(Mandatory, ParameterSetName = "TotalCIM")]
        [Parameter(Mandatory, ParameterSetName = "UsedCIM")]
        [Microsoft.Management.Infrastructure.CimSession[]]$CimSession,

        [Parameter(HelpMessage = "Enter the minimum % free memory", ParameterSetName = "PercentComputer")]
        [Parameter(HelpMessage = "Enter the minimum % free memory", ParameterSetName = "PercentCIM")]
        [ValidateNotNullorEmpty()]
        [int]$PercentFree = 50,

        [Parameter(HelpMessage = "Enter the minimum free memory in GB", Mandatory, ParameterSetName = "FreeComputer")]
        [Parameter(HelpMessage = "Enter the minimum free memory in GB", Mandatory, ParameterSetName = "FreeCIM")]
        [ValidateNotNullorEmpty()]
        [double]$FreeGB,

        [ValidateNotNullorEmpty()]
        [Parameter(HelpMessage = "Enter the minimum total memory in GB", Mandatory, ParameterSetName = "TotalComputer")]
        [Parameter(HelpMessage = "Enter the minimum total memory in GB", Mandatory, ParameterSetName = "TotalCIM")]
        [int]$TotalGB,

        [Parameter(HelpMessage = "Enter the minimum amount of used memory in GB", Mandatory, ParameterSetName = "UsedComputer")]
        [Parameter(HelpMessage = "Enter the minimum amount of used memory in GB", Mandatory, ParameterSetName = "UsedCIM")]
        [ValidateNotNullorEmpty()]
        [double]$UsedGB,

        [switch]$Quiet
    )

    Begin {
        Write-Verbose "[BEGIN  ] Starting: $($MyInvocation.Mycommand)"
    } #begin

    Process {
        Write-Verbose "[PROCESS] PSBoundParameters"
        Write-Verbose ($PSBoundParameters | Out-String)

        if ($Computername) {
            $usage = foreach ($Computer in $Computername) {
                #get memory usage data for each computer
                Get-MemoryUsage -Computername $computer
            }
        }
        else {
            $usage = foreach ($session in $CIMSession) {
                #get memory usage data for each computer
                Get-MemoryUsage -CimSession $session
            }
        }

        Foreach ($mem in $usage) {

            Switch -regex ($PSCmdlet.ParameterSetName) {
                "Used" {
                    Write-Verbose "[PROCESS] Testing if Used GB is >= to $UsedGB"
                    $used = $mem.TotalGB - $mem.FreeGB
                    Write-Verbose "[PROCESS] Used = $used"
                    if ($Used -ge $usedGB) {
                        $Test = $True
                    }
                    else {
                        $Test = $False
                    }
                    $data = $mem | Select-Object Computername, @{Name = "UsedGB"; Expression = { $used } },
                    @{Name = "Test"; Expression = { $test } }
                }
                "Total" {
                    Write-Verbose "[PROCESS] Testing if Total size is >= $TotalGB"
                    if ($mem.TotalGB -ge $TotalGB) {
                        $Test = $True
                    }
                    else {
                        $Test = $False
                    }
                    $data = $mem | Select-Object Computername, TotalGB, @{Name = "Test"; Expression = { $test } }
                }
                "Free" {
                    Write-Verbose "[PROCESS] Testing if Free GB is >= $FreeGB"
                    if ($FreeGB -le $mem.FreeGB) {
                        $Test = $True
                    }
                    else {
                        $Test = $False
                    }
                    $data = $mem | Select-Object Computername, FreeGB, @{Name = "Test"; Expression = { $test } }
                }
                "Percent" {
                    Write-Verbose "[PROCESS] Testing if Percent free is >= $PercentFree"
                    if ($mem.PctFree -ge $percentFree) {
                        $Test = $True
                    }
                    else {
                        $Test = $False
                    }
                    $data = $mem | Select-Object Computername, PctFree, @{Name = "Test"; Expression = { $test } }
                }
            } #switch

            if ($Quiet) {
                $Test
            }
            else {
                $data
            }
        } #foreach $mem

    } #process
    End {
        Write-Verbose "[END    ] Ending: $($MyInvocation.Mycommand)"
    } #end

} #end Test-MemoryUsage

Function Get-MemoryPerformance {

    [cmdletbinding(DefaultParameterSetName = "Computername")]
    [alias("gmemp")]
    [OutputType("Selected.Microsoft.Management.Infrastructure.CimInstance")]
    Param(
        [Parameter(
            Position = 0,
            ValueFromPipeline,
            ValueFromPipelineByPropertyName,
            ParameterSetName = 'Computername'
        )]
        [ValidateNotNullorEmpty()]
        [Alias("cn")]
        [string[]]$Computername = $env:Computername,

        [Parameter(
            ParameterSetName = 'Cim',
            Mandatory,
            ValueFromPipeline
        )]
        [ValidateNotNullorEmpty()]
        [Microsoft.Management.Infrastructure.CimSession[]]$CimSession
    )

    Begin {
        Write-Verbose "[BEGIN  ] Starting: $($MyInvocation.Mycommand)"
        <#
        Get all memory performance counters. Assuming counters on the
        client are the same as on the server. Sort by name.
        #>
        $all = (Get-Counter -ListSet Memory*).counter | Sort-Object

        #get a list of class properties. Some of the properties don't
        #appear to have any values and are different than what you get
        #with Get-Counter
        $perfclass = Get-CimClass -ClassName Win32_PerfFormattedData_PerfOS_Memory
        $selected = $perfclass.CimClassProperties | Select-Object -Skip 9 -ExpandProperty Name
        $selected += @{Name = "DateTime"; Expression = { (Get-Date) } }
        $selected += @{Name = "ComputerName"; Expression = { $session.ComputerName } }
        Write-Verbose "[BEGIN  ] PSBoundParameters"
        Write-Verbose ($PSBoundParameters | Out-String)
    } #begin

    Process {

        Write-Verbose "[PROCESS] Using parameter set $($PSCmdlet.ParameterSetName)"

        if ($pscmdlet.ParameterSetName -eq 'Computername') {
            #create a temporary cimsession if using a computername
            $MyCIMSession = foreach ($item in $Computername) {
                Try {
                    Write-Verbose "[PROCESS] Creating temporary CIM Session to $item"
                    New-CimSession -ComputerName $item -ErrorAction Stop -OutVariable +tmpSess
                    Write-Verbose "[PROCESS] Added session"
                }
                Catch {
                    Write-Error "[$($item.toUpper())] Failed to create temporary CIM Session. $($_.exception.message)"
                }
            } #foreach item in computername
        } #if computername parameter set
        else {
            Write-Verbose "[PROCESS] Re-using CimSessions"
            $MyCimSession = $CimSession
        }

        foreach ($session in $MyCIMSession) {
            Try {
                Get-CimInstance -ClassName Win32_PerfFormattedData_PerfOS_Memory -CimSession $session |
                    Select-Object -Property $selected
            } #try
            Catch {
                Write-Error "Failed to get performance data from $($session.computername.toupper()). $($_.exception.message)"
            }
        } #foreach

        #clean up
        if ($tmpSess) {
            Write-Verbose "[PROCESS] Removing temporary sessions"
            $tmpSess | Remove-CimSession
            Remove-Variable tmpsess
        }

    } #process

    End {
        Write-Verbose "[END    ] Ending: $($MyInvocation.Mycommand)"
    } #end
} #end Get-MemoryPerformance
Function Get-PhysicalMemory {

    [cmdletbinding(DefaultParameterSetName = "Computername")]
    [alias("gpmem")]
    [OutputType("physicalMemoryUnit")]
    Param(
        [Parameter(
            Position = 0,
            ValueFromPipeline,
            ValueFromPipelineByPropertyName,
            ParameterSetName = 'Computername'
        )]
        [ValidateNotNullorEmpty()]
        [Alias("cn")]
        [string[]]$Computername = $env:Computername,

        [Parameter(
            ParameterSetName = 'Cim',
            Mandatory,
            ValueFromPipeline
        )]
        [ValidateNotNullorEmpty()]
        [Microsoft.Management.Infrastructure.CimSession[]]$CimSession
    )

    Begin {
        Write-Verbose "[BEGIN  ] Starting: $($MyInvocation.Mycommand)"
        Write-Verbose "[BEGIN  ] PSBoundParameters"
        Write-Verbose ($PSBoundParameters | Out-String)

        #define a hash table to resolve Form factor
        $form = @{
            0  = 'Unknown'
            1  = 'Other'
            2  = 'SIP'
            3  = 'DIP'
            4  = 'ZIP'
            5  = 'SOJ'
            6  = 'Proprietary'
            7  = 'SIMM'
            8  = 'DIMM'
            9  = 'TSOP'
            10 = 'PGA'
            11 = 'RIMM'
            12 = 'SODIMM'
            13 = 'SRIMM'
            14 = 'SMD'
            15 = 'SSMP'
            16 = 'QFP'
            17 = 'TQFP'
            18 = 'SOIC'
            19 = 'LCC'
            20 = 'PLCC'
            21 = 'BGA'
            22 = 'FPBGA'
            23 = 'LGA'
        }

    } #begin

    Process {

        Write-Verbose "[PROCESS] Using parameter set $($PSCmdlet.ParameterSetName)"

        if ($pscmdlet.ParameterSetName -eq 'Computername') {
            #create a temporary cimsession if using a computername
            $MyCIMSession = foreach ($item in $Computername) {
                Try {
                    Write-Verbose "[PROCESS] Creating temporary CIM Session to $item"
                    New-CimSession -ComputerName $item -ErrorAction Stop -OutVariable +tmpSess
                    Write-Verbose "[PROCESS] Added session"
                }
                Catch {
                    Write-Error "[$($item.toUpper())] Failed to create temporary CIM Session. $($_.exception.message)"
                }
            } #foreach item in computername
        } #if computername parameter set
        else {
            Write-Verbose "[PROCESS] Re-using CimSessions"
            $MyCimSession = $CimSession
        }

        foreach ($session in $MyCIMSession) {
            Write-Verbose "[PROCESS] Processing $($session.computername)"
            Try {
                $data = Get-CimInstance -ClassName win32_physicalmemory -CimSession $session -ErrorAction Stop

            } #Try
            Catch {
                Write-Error "[$($Session.ComputerName.toUpper())] $($_.exception.message)"
            }
            if ($data) {
                #create a custom object
                foreach ($item in $data) {
                    [pscustomobject]@{
                        PSTypeName   = "physicalMemoryUnit"
                        Computername = $item.PSComputername.ToUpper()
                        Manufacturer = $item.Manufacturer
                        CapacityGB   = $item.Capacity / 1GB
                        Form         = $form.item($item.FormFactor -as [int])
                        ClockSpeed   = $item.ConfiguredClockSpeed
                        Voltage      = $item.ConfiguredVoltage
                        Location     = $item.DeviceLocator
                    }
                }
            }
        } #foreach
        #clean up
        if ($tmpSess) {
            Write-Verbose "[PROCESS] Removing temporary sessions"
            $tmpSess | Remove-CimSession
            Remove-Variable tmpsess
        }
    } #process

    End {
        Write-Verbose "[END    ] Ending: $($MyInvocation.Mycommand)"
    } #end

} #get-PhysicalMemory

Function Get-TopProcessMemory {

    [cmdletbinding(DefaultParameterSetName = "Computername")]
    [alias("gtop")]
    [OutputType("topProcessMemoryUnit")]
    Param(
        [Parameter(
            Position = 0,
            ValueFromPipeline,
            ValueFromPipelineByPropertyName,
            ParameterSetName = 'Computername'
        )]
        [ValidateNotNullorEmpty()]
        [Alias("cn")]
        [string[]]$Computername = $env:Computername,

        [Parameter(
            ParameterSetName = 'Cim',
            Mandatory,
            ValueFromPipeline
        )]
        [ValidateNotNullorEmpty()]
        [Microsoft.Management.Infrastructure.CimSession[]]$CimSession,

        [validateRange(1, 25)]
        [int]$Top = 5
    )

    Begin {
        Write-Verbose "[BEGIN  ] Starting: $($MyInvocation.Mycommand)"

        #a private function to get process owner
        function _getProcessOwner {
            [cmdletbinding()]
            Param([object]$CimProcess)
            $own = $cimProcess | Invoke-CimMethod -MethodName GetOwner
            "$($own.domain)\$($own.user)"
        }
    } #Begin

    Process {
        Write-Verbose "[PROCESS] PSBoundParameters"
        Write-Verbose ($PSBoundParameters | Out-String)

        Write-Verbose "[PROCESS] Using parameter set $($PSCmdlet.ParameterSetName)"

        if ($pscmdlet.ParameterSetName -eq 'Computername') {
            #create a temporary cimsession if using a computername
           $myCIMSession = foreach ($item in $Computername) {
                Try {
                    Write-Verbose "[PROCESS] Creating temporary CIM Session to $item"
                    New-CimSession -ComputerName $item -ErrorAction Stop -OutVariable +tmpSess
                    Write-Verbose "[PROCESS] Added session"
                }
                Catch {
                    Write-Error "[$($item.toUpper())] Failed to create temporary CIM Session. $($_.exception.message)"
                }
            } #foreach item in computername
        } #if computername parameter set
        else {
            Write-Verbose "[PROCESS] Re-using CimSessions"
            $MyCimSession = $CimSession
        }

        foreach ($session in $MyCIMSession) {
            Write-Verbose "[PROCESS] SessionID = $($session.ID)"
            Try {
                Write-Verbose "[PROCESS] Querying $($session.computername.ToUpper())"
                $data = Get-CimInstance -ClassName win32_process -CimSession $session -ErrorAction Stop | Sort-Object WorkingSetSize -Descending | Select-Object -First $Top
            } #Try
            Catch {
                Write-Error "[$($Session.Computername.toUpper())] $($_.exception.message)"
            } #Catch
            if ($data) {
                #get overall memory usage
                $mu = Get-MemoryUsage -CimSession $session
                [int64]$used = ($mu.totalGB - $mu.FreeGB) * 1gb

                Foreach ($item in $data) {
                    [PSCustomObject]@{
                        PSTypename   = "topProcessMemoryUnit"
                        Computername = $item.PSComputername.ToUpper()
                        ProcessID    = $item.ProcessID
                        Name         = $item.Name
                        WS           = $item.WorkingSetSize
                        PctUsed      = [math]::Round(($item.workingsetsize /$used) * 100, 4)
                        CreationDate = $item.CreationDate
                        Runtime      = (Get-Date) - $item.CreationDate
                        Owner        = $(_getProcessOwner $item)
                        Commandline  = $item.Commandline
                    }
                }
            }
        } #foreach
        #clean up
        if ($tmpSess.count -gt 0) {
            Write-Verbose "[PROCESS] Removing temporary sessions"
            $tmpSess | out-string | Write-Verbose
            $tmpSess | Remove-CimSession
            Remove-Variable -name tmpSess -Force
        }
    } #process

    End {
        Write-Verbose "[END    ] Ending: $($MyInvocation.Mycommand)"
    } #end
} #end function

